src/KubeOps.Abstractions/Events/IEventPublisherFactory.cs

```diff
@@ -0,0 +1,13 @@
+namespace KubeOps.Abstractions.Events;
+
+/// <summary>
+/// Represents a type used to create <see cref="EventPublisher"/>s for clients and controllers.
+/// </summary>
+public interface IEventPublisherFactory : IDisposable
```

Comment:
What is the reason for this factory?

Reply:
Exactly. For example, in my integration tests, I'd like to keep the code as close to production as possible. Therefore I use the `IHost` of my program and just replace single parts of that.

For example, this is the factory/class fixture I use right now:

```csharp
using k8s;
using KubeOps.Abstractions.Builder;
using Microsoft.AspNetCore.Hosting;
using Microsoft.AspNetCore.Mvc.Testing;
using Microsoft.Extensions.DependencyInjection;
using Xunit.Abstractions;
// ...and some more internal ones.

public sealed class DefaultApplicationFactory(ITestOutputHelper sink)
    : WebApplicationFactory<Program>
{
    private readonly KubernetesFixture _fixture = new();

    protected override void ConfigureWebHost(IWebHostBuilder builder)
    {
        builder.ConfigureServices(services => services
            .Replace<IKubernetesNamespaceProvider>(_fixture.Provider)
            .Replace<IKubernetes>(_fixture.Client)
            .Replace<KubernetesClientConfiguration>(_fixture.Settings)
            .Replace(new OperatorSettings
            {
                Namespace = _fixture.Namespace,
                Name = "integration-testing",
            })
        );

        base.ConfigureWebHost(builder);
    }
}
```

This allows me to leave the remaining parts as is and actually test the behaviour instead of my mocks. By adding additional factories, I can now do stuff like this:

```csharp
internal sealed class XunitEventPublisherFactory(ITestOutputHelper outputHelper) : IEventPublisherFactory
{
    public void Dispose() { }

    public EventPublisher Create() => (entity, reason, message, type) =>
    {
        outputHelper.WriteLine(
            $"Kubernetes Event for {entity.GetType().Name} published. Reason: {reason}\t Message: {message}");
        return Task.CompletedTask;
    };
}
```

Or I could check the invocations like you did with the `InvocationCounter` in the tests. It's not strictly necessary to add it, as the current `Func` also satisfies the same behavior, rather it's a nice addition. Also, if we want to enhance the event building mechanism in the future, e.g. with Prometheus metrics, this could also be done more easily as the factory allows us to use dependency injection.